tests/phpunit/integration/Core/Permissions/PermissionsTest.php

```diff
@@ -0,0 +1,206 @@
+<?php
+
+namespace Google\Site_Kit\Core\Permissions;
+
+use Google\Site_Kit\Context;
+use Google\Site_Kit\Core\Authentication\Authentication;
+use Google\Site_Kit\Core\Storage\Options;
+use Google\Site_Kit\Core\Storage\User_Options;
+use Google\Site_Kit\Tests\TestCase;
+
+class PermissionsTest extends TestCase {
+
+	public function test_register_without_dynamic_capabilities() {
+		$filters = array(
+			'map_meta_cap',
+			'googlesitekit_user_data',
+			'user_has_cap',
+		);
+
+		foreach ( $filters as $filter ) {
+			remove_all_filters( $filter );
+		}
+
+		$this->get_instance()->register();
+
+		foreach ( $filters as $filter ) {
+			$this->assertTrue( has_filter( $filter ) );
+		}
+	}
+
+	/**
+	 * @processIsolation
+	 */
+	public function test_register_with_dynamic_capabilities_enabled() {
+		$filters = array(
+			'map_meta_cap',
+			'googlesitekit_user_data',
+			'user_has_cap',
+		);
+
+		foreach ( $filters as $filter ) {
+			remove_all_filters( $filter );
+		}
+
+		define( 'GOOGLESITEKIT_DISABLE_DYNAMIC_CAPABILITIES', true );
+		$this->get_instance()->register();
+
+		$this->assertTrue( has_filter( 'map_meta_cap' ) );
+		$this->assertTrue( has_filter( 'googlesitekit_user_data' ) );
+		$this->assertFalse( has_filter( 'user_has_cap' ) );
+	}
+
+	/**
+	 * @dataProvider provider_role_without_permissions
+	 */
+	public function test_all_roles_without_setup_does_not_have_any_capability_attached_to_it( $role ) {
+		$user = self::factory()->user->create_and_get( array( 'role' => $role ) );
+		wp_set_current_user( $user->ID );
+		do_action( 'wp_login', $user->user_login, $user );
+
+		$this->get_instance()->register();
+
+		$this->assertTrue( has_filter( 'map_meta_cap' ) );
+		$this->assertTrue( has_filter( 'googlesitekit_user_data' ) );
+		$this->assertTrue( has_filter( 'user_has_cap' ) );
+
+		foreach ( Permissions::get_capabilities() as $capability ) {
+			$this->assertFalse( $user->has_cap( $capability ), "{$role} can {$capability}" );
+		}
+	}
+
+	public function provider_role_without_permissions() {
+		yield 'user with `subscriber` role' => array( 'subscriber' );
+		yield 'user with `contributor` role' => array( 'contributor' );
+		yield 'user with `author` role' => array( 'author' );
+		yield 'user with `editor` role' => array( 'editor' );
+	}
+
+	public function test_unauthenticated_user_with_administrator_role_without_setup() {
+		$user = self::factory()->user->create_and_get( array( 'role' => 'administrator' ) );
+		wp_set_current_user( $user->ID );
+		do_action( 'wp_login', $user->user_login, $user );
```

Comment:
This was mostly as I was facing some unexpected results but I think ti can be removed at this point, otherwise moving to `setUp` seems like a better idea.